Initial contribution of LLJB to the Eclipse OMR Project

nbhuiyan opened the pull request:
The LLJB project enables interoperability between LLVM and JitBuilder.
LLJB takes LLVM Intermediate Representation (IR) files and uses
JitBuilder API to construct equivalent OMR Compiler IL. The OMR Compiler
then performs optimizations and genates native code for every function
in the module.

LLJB is being contributed as a subproject of JitBuilder.

Signed-off-by: Nazim Uddin Bhuiyan <nazimudd@ualberta.ca>

0xdaryl commented on jitbuilder/lljb/include/lljb/Compiler.hpp:
```diff
@@ -0,0 +1,120 @@
+/*******************************************************************************
+ * Copyright (c) 2019, 2019 IBM Corp. and others
```
Why does the include directory have `lljb` as a subdirectory (since it already lives under a `lljb` directory)?  Sorry, I think I asked you this before a few months back but I forgot...

nbhuiyan replied:
It's just one way that many C++ projects (including llvm and clang) structure header files. This way, when specifying header directory paths for a project, you need to just add `project/include` folder without worrying about the order you are adding the paths. When including header files, you can be specific about exactly which header file you are using by specifying `include "lljb/MethodBuilder"` in your code file without having to worry about how the header search paths are ordered by the build system, removing the potential for including some other project's `MethodBuilder` file. Since some lljb's files have shared names with what exists in OMR already, it was simpler to structure the include directory this way.

nbhuiyan commented on cmake/config.cmake:
```diff
@@ -44,21 +44,36 @@ set(OMR_THREAD ON CACHE BOOL "Enable thread library")
 	if(OMR_TEST_COMPILER)
 		message(FATAL_ERROR "OMR_TEST_COMPILER is enabled but OMR_COMPILER is not enabled")
 	endif()
+	if(OMR_LLJB)
```
Actually, this line would not be reached because the check on L58 would cause an abort due to enabling JitBuilder but not the Compiler. Check on L60 would cause an abort when both Compiler and LLJB is enabled but JitBuilder is disabled. I'll remove that check.

nbhuiyan commented on fvtest/lljbtest/mandelbrot.cpp:
```diff
@@ -0,0 +1,99 @@
+    clockid_t clockid = CLOCK_MONOTONIC;
+
+    struct timespec start_time, end_time;
+    clock_gettime(clockid, &start_time);
```
I added the guard on the CMake file that handles generating IR from it.

0xdaryl commented:
@genie-omr build all

nbhuiyan commented:
@mstoodle I'd appreciate a review from you as well!

0xdaryl commented:
@genie-omr build zos

mstoodle commented on jitbuilder/lljb/src/IRVisitor.cpp:
```diff
@@ -0,0 +1,559 @@
+      TR::IlValue * ilSrc = getIlValue(source);
+      loadedVal = _builder->LoadAt(
+                              _td->PointerTo(_methodBuilder->getIlType(_td,I.getType())),
+                              _builder->
```
does this `IndexAt` do anything? I would have expected this whole `LoadAt` could just take i1Src as its operand.

nbhuiyan replied:
You are right. Using `IndexAt` was an extra step required during the early days of the project where local variables were being allocated as single-element arrays and the `ilSrc` was the `IlValue` representing the array. I got rid of this as it is now redundant.

mstoodle commented on jitbuilder/lljb/src/IRVisitor.cpp:
```diff
@@ -0,0 +1,559 @@
+   if (_methodBuilder->isIndirectLoadOrStore(dest))
+      {
+      _builder->StoreAt(
+               _builder->IndexAt(
```
Same "IndexAt" comment here...if the index is 0, I wouldn't have expected you need it.

nbhuiyan replied:
See reply to the other comment.

mstoodle commented on jitbuilder/lljb/src/IRVisitor.cpp:
```diff
@@ -0,0 +1,559 @@
+      {
+      case llvm::CmpInst::Predicate::ICMP_EQ:
+      case llvm::CmpInst::Predicate::FCMP_OEQ:
+      case llvm::CmpInst::Predicate::FCMP_UEQ:
```
We should really create Unordered comparison operations in JitBuilder that leverage the unordered fcmp opcodes in the OMR compiler rather than map both ordered and unordered fp compares to the same JitBuilder comparisons. Any code that's senstiive to unordered floating point values is likely to behave incorrectly here (at least if a NaN shows up), so maybe it's better to assert on those opcodes until the appropriate operations are added to JitBuilder?

nbhuiyan replied:
Done. I have also changed one of the test cases that used to generate `FCMP_UEQ` LLVM instruction.

mstoodle commented on jitbuilder/lljb/src/IRVisitor.cpp:
```diff
@@ -0,0 +1,559 @@
+      TR::IlBuilder * ifTrue = _methodBuilder->getByteCodeBuilder(I.getSuccessor(0));
+      TR::IlBuilder * ifFalse = _methodBuilder->getByteCodeBuilder(I.getSuccessor(1));
+      assert(ifTrue && ifFalse && condition && "Failed to find destination blocks for ifThenElse");
+      _builder->IfThenElse(&ifTrue, &ifFalse, condition);
```
I do not think this is what you want, although you might be getting away with it by fluke. I think you would be better off doing:
```
_builder->IfCmpEqualTo(&ifTrue, _builder->NotEqual(condition, _builder->Const*(0)));
_builder->Goto(&ifiFalse);
```

(not sure what kind of Const you need).

nbhuiyan replied:
Makes sense, thanks for pointing this out. I have incorporated your suggestion here.

mstoodle commented on jitbuilder/lljb/src/IRVisitor.cpp:
```diff
@@ -0,0 +1,559 @@
+               _builder->IndexAt(
+                  _td->PointerTo(
+                     _methodBuilder->getIlType(_td,I.getSourceElementType()->getArrayElementType())),
+                  getIlValue(I.getOperand(0)),
```
another `IndexAt` with index 0

nbhuiyan replied:
This is not `IndexAt` at index 0 similar to the Load/Store instruction visitors. `getelementptr` instruction is used for address computations involving aggregate data types. In this particular line where we deal with single dimension arrays, the second argument provided to `IndexAt` is `getIlValue(I.getOperand(0))` which gets the `IlValue` corresponding to the array allocated in an earlier instruction. `getOperand(0)` on `getelementptr` instruction gets the `llvm::Value` mapped to the `IlValue` we want here.

mstoodle commented on jitbuilder/lljb/src/IRVisitor.cpp:
```diff
@@ -0,0 +1,559 @@
+   }
+
+void
+IRVisitor::visitPHINode(llvm::PHINode &I)
```
As a general comment (irony coming), I think you could improve on the number of comments in this file.

But I think this function in particular deserves some comments. I understand what it's doing, but not sure I understand why (probably due to lack of llvm familiarity).

nbhuiyan replied:
I agree. I have added *some* comments... hope that helps :)

rwy0717 commented on fvtest/lljbtest/CMakeLists.txt:
```diff
@@ -0,0 +1,97 @@
+find_program(CLANG_EXECUTABLE  NAMES clang)
+
+if (NOT CLANG_EXECUTABLE)
+    message(SEND_ERROR "Could not find clang")
```
If it were me, I'd have used `FATAL_ERROR` instead. PS, you might want to factor this code out into a `FindClang.cmake` module. And, you might want to use the current compiler as a hint.

nbhuiyan replied:
I have changed this to `FATAL_ERROR` as suggested. As for factoring this out to a separate module, I would like to try that in a follow up PR instead of this one.

rwy0717 commented on fvtest/lljbtest/CMakeLists.txt:
```diff
@@ -0,0 +1,97 @@
+# [2] http://openjdk.java.net/legal/assembly-exception.html
+#
+# SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+###############################################################################
```
This whole file is using a mix of tabs and spaces.

nbhuiyan replied:
Fixed.

rwy0717 commented on fvtest/lljbtest/loops.cpp:
```diff
@@ -0,0 +1,39 @@
+   int y = 0;
+   int i = 0;
+
+   while (i < 10 && (x >= y || y >= 0)){ // this generates phi nodes
+      x += 1;
+      i++;
+   }
```
```suggestion
   while (i < 10 && (x >= y || y >= 0))
      {
      // this generates phi nodes
      x += 1;
      i++;
      }
```

nbhuiyan replied:
I have made the suggested changes.

rwy0717 commented on fvtest/lljbtest/loops.cpp:
```diff
@@ -0,0 +1,39 @@
+      i++;
+   }
+
+   for (i = 0; i < 7; i++){
+      y += 1;
+   }
```
```suggestion
   for (i = 0; i < 7; i++)
      {
      y += 1;
      }
```

nbhuiyan replied:
I have made the suggested changes.

rwy0717 commented on jitbuilder/lljb/CMakeLists.txt:
```diff
@@ -0,0 +1,55 @@
+# [2] http://openjdk.java.net/legal/assembly-exception.html
+#
+# SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+#############################################################################
```
File is using spaces, should be tabs.

nbhuiyan replied:
Fixed.

rwy0717 commented on cmake/config.cmake:
```diff
@@ -44,23 +44,35 @@ set(OMR_THREAD ON CACHE BOOL "Enable thread library")
+	if(NOT OMR_JITBUILDER)
+		message(FATAL_ERROR "OMR_LLJB is enabled but OMR_JITBUILDER is not enabled")
+	endif()
+	set (OMR_LLJB_TEST ON CACHE BOOL "Enable lljb tests")
```
```suggestion
	set(OMR_LLJB_TEST ON CACHE BOOL "Enable lljb tests")
```

nbhuiyan replied:
Fixed.

rwy0717 commented on fvtest/lljbtest/CMakeLists.txt:
```diff
@@ -0,0 +1,97 @@
+
+endfunction(generate_module_from_cpp)
+
+
```
bit of bonus whitespace

nbhuiyan replied:
Fixed.

rwy0717 commented on jitbuilder/lljb/CMakeLists.txt:
```diff
@@ -0,0 +1,55 @@
+    src/IRVisitor.cpp
+)
+
+get_property(JITBUILDER_INCLUDE TARGET jitbuilder PROPERTY INTERFACE_INCLUDE_DIRECTORIES)
```
Is this needed? lljb links against jitbuilder a few lines below, won't that give us the same thing?

nbhuiyan replied:
You are right, this is not needed. I thought I needed to get the `JITBUILDER_INCLUDE` property and explicitly needed to add them to the header search paths.

rwy0717 commented on jitbuilder/lljb/CMakeLists.txt:
```diff
@@ -0,0 +1,55 @@
+# SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+#############################################################################
+
+if(${CMAKE_SYSTEM_NAME} MATCHES "Darwin")
+    set(ENV{PATH} "/usr/local/opt/llvm\@8/bin/:${PATH}")
+endif()
```
Is this meant to serve as a guess for find_package(LLVM) below? If so, would it be better to just hand this path to find_package, using the PATHS option?

nbhuiyan replied:
Yup, I have changed this to provide this path to `find_package` directly. It was not clear to me in the documentation, but if I do specify a `PATHS` value, are the paths in $PATH also searched as a fallback? If that's not the case, if llvm version 8 is the default and the only installation a macOS environment, CMake will not find LLVMConfig.cmake.

rwy0717 commented on jitbuilder/lljb/CMakeLists.txt:
```diff
@@ -0,0 +1,55 @@
+    PUBLIC
+        include/
+        ${JITBUILDER_INCLUDE}
+        ${CMAKE_CURRENT_SOURCE_DIR}/../../compiler
+        ${CMAKE_CURRENT_SOURCE_DIR}/../..
```
I would express this as `${omr_SOURCE_DIR}/compiler`. Isn't `../..` just the root of OMR? You probably don't need that.

rwy0717 commented on fvtest/lljbtest/CMakeLists.txt:
```diff
@@ -0,0 +1,97 @@
+
+target_include_directories(lljb_run
+    PRIVATE
+        ${CMAKE_SOURCE_DIR}/jitbuilder/
+        ${CMAKE_SOURCE_DIR}/compiler/
+        ${CMAKE_SOURCE_DIR}/omr/
```
These three paths should be rooted by `${omr_SOURCE_DIR}`, so that the paths are still correct when OMR is a subdirectory of a bigger project.

rwy0717 replied:
Actually, these shouldn't be needed, since lljb should provide all necessary include directories.

rwy0717 commented on fvtest/lljbtest/lljb_run.cpp:
```diff
@@ -0,0 +1,57 @@
+   bool jitInitialized = initializeJit();
+   assert(jitInitialized && "Jit Failed to initialize");
+
+   const char * filename = argv[1];
```
I think this program ought to check argc before proceeding.

rwy0717 commented on fvtest/lljbtest/CMakeLists.txt:
```diff
@@ -0,0 +1,97 @@
+    LLVMIRReader
+    LLVMCore
+    LLVMSupport
+	omrGtest
```
I must be wrong but, is lljb_run actually using omrgtest? It looks like it's not.

rwy0717 replied:
Actually, lljb provides LLVMIRRReader, LLVMCore, and LLVMSupport through it's interface, so there's no need to link against those explicitly either.

nbhuiyan replied:
Fixed.

rwy0717 commented on fvtest/lljbtest/lljb_run.cpp:
```diff
@@ -0,0 +1,57 @@
+#include <iostream>
+#include <cassert>
+
+int main(int argc, char * argv[])
```
For portability, you might consider using `omr_main_function`, instead of directly writing your own main.

rwy0717 replied:
https://github.com/eclipse/omr/tree/master/util/main_function

rwy0717 replied:
I'm going to mark this conversation as resolved, as I think it can be addressed in a later PR (if ever).

rwy0717 commented on fvtest/lljbtest/lljb_run.cpp:
```diff
@@ -0,0 +1,57 @@
+   {
+
+   bool jitInitialized = initializeJit();
+   assert(jitInitialized && "Jit Failed to initialize");
```
This assert will become a no-op in release mode. You should replace this macro with if/print/exit.

rwy0717 commented on fvtest/lljbtest/CMakeLists.txt:
```diff
@@ -0,0 +1,97 @@
+    add_custom_target(generate_module_from_cpp_${src} ALL
+        DEPENDS ${CMAKE_CURRENT_BINARY_DIR}/${src}.ll
+    )
+
```
```suggestion
```

rwy0717 commented on fvtest/lljbtest/CMakeLists.txt:
```diff
@@ -0,0 +1,97 @@
+find_package(LLVM 8 REQUIRED CONFIG)
+
+# we need clang to generate IR from C++
+find_program(CLANG_EXECUTABLE  NAMES clang)
```
Another tiny nitpick, you're using clang to compile c++ code, should this be clang++?

nbhuiyan replied:
I have changed this to clang++. Both clang and clang++ seems to be interchangeable in my experience.

rwy0717 commented on fvtest/lljbtest/CMakeLists.txt:
```diff
@@ -0,0 +1,97 @@
+endif()
+
+# Generate llvm bitcode files from C++ using clang
+function(generate_module_from_cpp src)
```
Rather that call this function ..._from_cpp, I think from_cxx would be more conventional.

nbhuiyan commented:
@0xdaryl @mstoodle @rwy0717 thank you for taking the time to review such a large PR. I have addressed your suggestions.

To answer some of Mark's questions...
> Do the tests cover all the llvm opcodes you currently support?

Yes, every llvm opcodes that LLJB can handle are generated by the test cases.

> On what platforms have you tested LLJB so far?

I have built and tested LLJB on macOS, x86-64 Linux, and x86-64 Windows. However, building on Windows is quite complicated and long because LLVM currently does not include any LLVM developer libraries as part of any releases for Windows, requiring us to actually build them from source - which takes almost an eternity no matter how much I try to trim the build playing around with the configuration options. In the future, if LLVM allows us to easily download and install LLVM libraries on Windows similar to how easy it is to do that on macOS and Linux we can consider counting Windows as a supported platform for LLJB and making the necessary updates in the CMake files to build on windows.

nbhuiyan commented:
It is worth noting that LLJB is not currently built as part of any CI builds. This was due to the limitations posed by what was available to use before Azure Pipelines (i.e, Windows on AppVeyor and TravisCI). In an upcoming PR, I will enable building LLJB on macOS and Linux, where we can have a discussion around the impact of the additional CI build dependencies and build time introduced by LLJB.

rwy0717 commented on fvtest/lljbtest/lljb_run.cpp:
```diff
@@ -0,0 +1,61 @@
+
+int main(int argc, char * argv[])
+   {
+   if (argc != 2){
+      std::cerr << "Usage: ./lljb <module.ll>" << std::endl;
+      exit(EXIT_FAILURE);
+   }
+
+   if (!initializeJit()){
+      std::cerr << "Failed to initialize JIT" << std::endl;
+      exit(EXIT_FAILURE);
+   }
```
Whitespace nitpick, hope I'm not annoying yet :)

```suggestion
   if (argc != 2)
      {
      std::cerr << "Usage: ./lljb <module.ll>" << std::endl;
      exit(EXIT_FAILURE);
      }

   if (!initializeJit())
      {
      std::cerr << "Failed to initialize JIT" << std::endl;
      exit(EXIT_FAILURE);
      }
```

nbhuiyan replied:
Good catch once again! :)

0xdaryl commented:
@genie-omr build all

0xdaryl commented:
I will merge this tomorrow based on reviewer approvals and the resolution of contingent problems.  Last call for comments.
